Add additional Sentry E2E tests

Gudahtt opened the pull request:
## Explanation

Additional E2E tests have been added for our Sentry error handling.

Before this PR, the Sentry e2e tests before initialization only tested background errors, and the after initialization tests only tested UI errors. Now both types of errors are tested in both scenarios.

Tests have been added to test the state object sent along with each Sentry error as well. At the moment this object is empty in some circumstances, but this will change in later PRs.

To trigger background errors for testing, a new controller API method has been added that throws a test error. This mirrors a similar test function in the UI. A `@deprecated` tag was used to warn against using this accidentally in real code.

State snapshots have been added to make it easier to track changes to the Sentry state in future PRs. This will be very useful for the next few PRs, but might be less valuable long-term. We should consider removing it after the next few Sentry state changes to reduce maintenance burdens. State snapshots can be updated by setting the environment variable `UPDATE_SNAPSHOTS` to `true` when running tests, e.g. 

```
UPDATE_SNAPSHOTS=true yarn test:e2e:single --browser chrome ./test/e2e/tests/errors.spec.js
```

Relates to #20449

## Manual Testing Steps

N/A

## Pre-merge author checklist

- [x] I've clearly explained:
  - [x] What problem this PR is solving
  - [x] How this problem was solved
  - [x] How reviewers can test my changes
- [x] Sufficient automated test coverage has been added

## Pre-merge reviewer checklist

- [ ] Manual testing (e.g. pull and build branch, run in browser, test code being changed)
- [ ] PR is linked to the appropriate GitHub issue
- [ ] **IF** this PR fixes a bug in the release milestone, add this PR to the release milestone

If further QA is required (e.g. new feature, complex testing steps, large refactor), add the `Extension QA Board` label.

In this case, a QA Engineer approval will be be required.

Gudahtt committed:
Reorganize Sentry error e2e tests

The tests have been reorganized into different describe blocks. Each
describe block is for either before or after initialization, and either
with or without opting into metrics.

This was done to simplify later test additions. The conditions for each
test are now in the describe block, letting us test additional things
in each of these conditions. The conditions were flattened to a single
level to avoid excessive indentation.

Gudahtt committed:
Add error e2e test for background and UI errors

The Sentry e2e tests before initialization only tested background
errors, and the after initialization tests only tested UI errors. Now
both types of errors are tested in both scenarios.

Gudahtt committed:
Add error e2e tests for Sentry error state

E2E tests have been added to test the state object sent along with each
Sentry error.

At the moment this object is empty in some circumstances, but this will
change in later PRs.

danjm commented on app/scripts/metamask-controller.js:
```diff
@@ -4535,6 +4538,21 @@ export default class MetamaskController extends EventEmitter {
+   * @deprecated This is only mean to facilitiate E2E testing. We should not
+   * use this for handling errors.
+   */
+  throwError(message) {
```
nit: maybe make this a noop if we are not in dev or in test?

Gudahtt replied:
We could! If this was called in a production build it would legitimately be an error though. Maybe it'd be better to throw, in case someone doesn't heed the deprecation notice.

Though... I should at least skip setting up the hook in the UI in test builds. We don't want people calling that in the web console. I'll update that to only set it up on test or dev builds.

Gudahtt replied:
Revised here: https://github.com/MetaMask/metamask-extension/pull/20425/commits/f4f19321c24a28b5144556340f54a7a44071a1d8
Let me know if that makes sense

danjm replied:
yes, that makes sense

Gudahtt committed:
Rename throw test error function

Gudahtt committed:
Only setup debug/test state hooks in dev/test builds

The state hooks used for debugging and testing are now only included in
dev or test builds. The function name was updated and given a JSDoc
description to explain this more clearly as well.

Gudahtt commented on test/e2e/tests/errors.spec.js:
```diff
@@ -163,13 +455,15 @@ describe('Sentry errors', function () {
             const isPending = await mockedEndpoint.isPending();
             return isPending === false;
-          }, 10000);
+          }, 3000);
```
This was made 3000 to match the other test. If it takes longer than 3 seconds here for the request to arrive, then it implies the test where we're expecting it to never arrive isn't waiting long enough.

Gudahtt commented on test/e2e/tests/errors.spec.js:
```diff
@@ -133,9 +290,12 @@ describe('Sentry errors', function () {
           await driver.press('#password', driver.Key.ENTER);
+
           // Trigger error
-          driver.executeScript('window.stateHooks.throwTestError()');
```
These two lines were missing `await`s

Gudahtt committed:
Add state snapshot assertions

State snapshot assertions have been added to the e2e error tests. These
snapshots will be very useful in reviewing a few PRs that will follow
this one.

We might decide to remove these snapshots after this set of Sentry
refactors, as they might be more work to maintain than they're worth.
But they will be useful at least in the short-term.

The login step has been removed from a few tests because it introduced
indeterminacy (the login process continued asynchronously after the
login, and sometimes was not finished when the error was triggered).

Gudahtt commented on test/e2e/tests/errors.spec.js:
```diff
@@ -131,11 +383,10 @@ describe('Sentry errors', function () {
         },
         async ({ driver, mockedEndpoint }) => {
           await driver.navigate();
-          await driver.fill('#password', 'correct horse battery staple');
```
The login step was producing indeterminacy in the state snapshots because we didn't wait for the login process to finish. Rather than adding a waiting step, the login was removed (it wasn't necessary for the test).

Gudahtt committed:
Ensure login page has rendered during setup

This fixes an intermittent failure on Firefox

Gudahtt committed:
Format snapshots with prettier before writing them

Gudahtt committed:
Use defined set of date fields rather than infering from name

Gudahtt committed:
Remove waits for error screen

The error screen only appears after a long timeout, and it doesn't
affect the next test steps at all.

danjm commented on test/e2e/tests/errors.spec.js:
```diff
@@ -1,8 +1,70 @@
+  for (const field of dateFields) {
+    if (has(data, field)) {
+      set(data, field, typeof get(data, field));
+    }
```
sleek lodash usage :sunglasses:

danjm commented on test/e2e/tests/errors.spec.js:
```diff
@@ -1,8 +1,70 @@
+      return;
+    }
+    throw error;
+  }
```
I was a bit confused by this function, wondering why we don't communicate anything about the `deepStrictEqual` failure and why we just update the snapshot instead of failing the test. But then I re-read the PR description and the default value of the `update` param and this makes sense. We do throw there error except in cases where `UPDATE_SNAPSHOTS` has been explicitly set, which would be done intentionally when the snapsshots are to be updated.

:+1:

danjm commented:
Looks good to me. Nice improvements to the sentry error tests.

darkwing commented on test/e2e/tests/errors.spec.js:
```diff
@@ -73,7 +135,43 @@ describe('Sentry errors', function () {
+          );
+
+          // Wait for Sentry request
+          await driver.delay(3000);
```
Non-blocking nit but we may want to stick to known `delay` constants rather than an arbitrary number.  Can be done in a follow up though.

Gudahtt replied:
yeah good call. This was copied from pre-existing tests in this file, and they should all be using a matching delay time, so I'll leave this for a follow-up.

darkwing reviewed (approved):
Minor non-blocking nits re: `3000`, but otherwise good!
